### ISSUE TITLE
Newlines around comments getting mangled

### ISSUE THREAD

**Issue by gabrielbauman** [thymeleaf/thymeleaf#151]
In HTML5 mode using 2.0.15,

``` html
          </ul>
        </div>
        <!--/.nav-collapse -->
      </div>
```

becomes

``` html
          </ul>
        </div><!--/.nav-collapse -->

      </div>
```

This happens for every HTML comment. Not super critical but makes for some ugly HTML.

**Comment by gabrielbauman** [comment]
I should mention I am using the Spring3, Apache Tiles, and Conditional Comments dialects.
Here's some startup info:

```
* Cache Factory implementation: org.thymeleaf.cache.StandardCacheManager
* Template modes:
    * LEGACYHTML5
    * HTML5
    * XML
    * VALIDXHTML
    * XHTML
    * VALIDXML
* Template resolvers (in order):
    * [1] org.thymeleaf.templateresolver.ServletContextTemplateResolver
* Message resolvers (in order):
    * org.thymeleaf.spring3.messageresolver.SpringMessageResolver
* Dialect [1 of 3]: org.thymeleaf.spring3.dialect.SpringStandardDialect
    * Prefix: "th"
* Dialect [2 of 3]: org.thymeleaf.extras.tiles2.dialect.TilesDialect
    * Prefix: "tiles"
* Dialect [3 of 3]: org.thymeleaf.extras.conditionalcomments.dialect.ConditionalCommentsDialect
    * Prefix: "condcom"
* Leniencies by prefix:
    * "condcom": NON-LENIENT
    * "tiles": NON-LENIENT
    * "th": NON-LENIENT
TEMPLATE ENGINE CONFIGURED OK
```

**Comment by ultraq** [comment]
@danielfernandez I've always wanted to do something about these output formatting issues (a related issue is #113 ), but was wondering if that was something you'd want to target for the 2.0.x branch or for future versions of Thymeleaf?  (Note: I've already got a fix for #113 in my local copy of Thymeleaf and I've been running it on my website for a few months now.)

**Comment by danielfernandez** [comment]
@ultraq this looks like a good addition for 2.1. What solution are you planning to adopt? Off the top of my head, I think what could be done is repeating the whitespace node (containing any `\n`'s) after the element that is being repeated along with the repeated element itself...

**Comment by ultraq** [comment]
I think I did exactly that for the other issue, but for this one I'm not too sure how to go about it yet.  I'm mostly wondering why the whitespace before the comment node is lost, so wanted to delve into the parser to see what's going on there first before figuring out how to tackle this one.

**Comment by danielfernandez** [comment]
It might be the SAX parser itself not reporting the whitespace Text nodes around a Comment one. If so, we'd not have much to do until attoparser is adopted :-(... the only other possibility would be to process whitespace at the `org.thymeleaf.templateparser.TemplatePreprocessingReader` class, for example by surrounding it with some very uncommon unicode char. That way it wouldn't be considered whitespace and the parser would not ignore it...

...but, to be honest, I think that would be quite overkill. Every extra processing done at the `TemplateProcessingReader` goes directly against the template parsing speed figures :-(...

**Comment by ultraq** [comment]
I did some digging, and I think you're gonna like this :)  In the `AbstractNonValidatingSAXTemplateParser` file, `XmlSAXHandler` private class, you don't call `flushBuffer()` before creating a `Comment` node.  So I added a single call to `flushBuffer()` in there, and suddenly _ALL_ comments in the output from Thymeleaf magically readjusted themselves to the way they were in my original HTML files.

This was against version 2.0.15 (I should really update my workspace...) running in HTML5 template mode.  I haven't yet checked if I have to do a similar thing to the other parsers.

**Comment by danielfernandez** [comment]
Oh! great :-) Please fix both 2.0-dev and 2.1-dev. With a bit of luck, you will be able to just commit the fix on 2.0-dev and then cherry-pick the commit into 2.1-dev.

I'm therefore assigning this issue to milestone 2.0.17, and leave #113 for 2.1 :-)

**Comment by ultraq** [comment]
`flushBuffer()` fix committed in both 2.0-dev and 2.1-dev.
